plugins/modules/aws_msk_config.py

```diff
@@ -0,0 +1,304 @@
+        r_dict[kv[0].strip()] = kv[1].strip()
+    return r_dict
+    # python >= 2.7 is required:
+    # return {
+    #     k.strip(): v.strip() for k, v in (i.split("=") for i in p.decode().split("\n"))
+    # }
+
+
+@AWSRetry.backoff(tries=5, delay=5)
+def get_configurations_with_backoff(client):
+    paginator = client.get_paginator("list_configurations")
+    return paginator.paginate().build_full_result()
+
+
+def find_active_config(client, module):
+    """
+    looking for configuration by name
+      status is not returned for list_configurations in botocore 1.17.42
+      delete_configuration method was added in botocore 1.17.48
+    """
+
+    name = module.params["name"]
+
+    try:
+        all_configs = get_configurations_with_backoff(client)["Configurations"]
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="failed to obtain kafka configurations")
+
+    active_configs = list(
+        item
+        for item in all_configs
+        if item["Name"] == name and item["State"] == "ACTIVE"
+    )
+
+    if active_configs:
+        if len(active_configs) == 1:
+            return active_configs[0]
+        else:
+            module.fail_json_aws(
+                msg="found more than one active config with name '{0}'".format(name)
+            )
+
+    return None
+
+
+def create_config(client, module):
+    """create new or update existing configuration"""
+
+    config = find_active_config(client, module)
+
+    # create new configuration
+    if not config:
+
+        if module.check_mode:
+            return True, {}
+
+        try:
+            response = client.create_configuration(
+                Name=module.params.get("name"),
+                Description=module.params.get("description"),
+                KafkaVersions=module.params.get("kafka_versions"),
+                ServerProperties=dict_to_prop(module.params.get("config")).encode(),
+            )
+        except (
+            botocore.exceptions.BotoCoreError,
+            botocore.exceptions.ClientError,
+        ) as e:
+            module.fail_json_aws(e, "failed to create kafka configuration")
+
+    # update existing configuration (creates new revision)
+    else:
+        # it's required because 'config' doesn't contain 'ServerProperties'
+        response = client.describe_configuration_revision(
+            Arn=config["Arn"], Revision=config["LatestRevision"]["Revision"]
+        )
+
+        # compare configurations (description and properties) and update if required
+        # prop_module = {str(k): str(v) for k, v in module.params.get("config").items()}
+        prop_module = {}
+        for k, v in module.params.get("config").items():
+            prop_module[str(k)] = str(v)
+        if prop_to_dict(response.get("ServerProperties", "")) == prop_module:
+            if response.get("Description", "") == module.params.get("description"):
+                return False, response
```

Comment:
It's probably better to do the comparison in a dedicated function, that would mean you could simplify things a little:

```
def _is_changed(current, module):
    prop_module = {}
    for k, v in module.params.get("config").items():
        prop_module[str(k)] = str(v)
    if prop_to_dict(response.get("ServerProperties", "")) != prop_module:
        return True
    if response.get("Description", "") != module.params.get("description"):
        return True
    return False
```

```
        if not _is_changed(response, module):
            return False, response
```

it makes the code a little easier to follow

Reply:
agree, done, thanks for suggestion